torch/distributed/_sharded_tensor/api.py

```diff
@@ -667,21 +668,21 @@ def __setstate__(self, state):
 def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorInitParams):
     """ Helper to construct tensor from size, device and common params. """
 
-    if tensor_init_params.create_op == CreateOp.ONES:
-        return torch.ones(*size,
-                          dtype=tensor_init_params.dtype,
-                          layout=tensor_init_params.layout,
-                          device=local_device,
-                          pin_memory=tensor_init_params.pin_memory,
-                          requires_grad=tensor_init_params.requires_grad,)
-    elif tensor_init_params.create_op == CreateOp.EMPTY:
-        return torch.empty(*size,
-                           dtype=tensor_init_params.dtype,
-                           layout=tensor_init_params.layout,
-                           device=local_device,
-                           requires_grad=tensor_init_params.requires_grad,
-                           # Note memory_format param is not accepted by torch.ones
-                           memory_format=tensor_init_params.memory_format,
-                           pin_memory=tensor_init_params.pin_memory,)
+    create_op = tensor_init_params.create_op
+    dtype = tensor_init_params.tensor_properties.dtype
+    layout = tensor_init_params.tensor_properties.layout
+    requires_grad = tensor_init_params.tensor_properties.requires_grad
+    memory_format = tensor_init_params.tensor_properties.memory_format
+    pin_memory = tensor_init_params.tensor_properties.pin_memory
+
+    if create_op == CreateOp.ONES:
```

Comment:
@pritamdamania87 - why don't we just pass the function like `torch.ones` directly? that'd feel more pythonic. And should other args be just `kwargs` too? And generally, I wonder whether there's a similar concept to TensorInitParams somewhere in core already. There's definitely TensorOptions in C++, but I think it's just kwargs in python.

Reply:
@dzhulgakov That is a good point, I think we can just change `tensor_init_params` to something like `tensor_creator` and make it a callable that returns a torch.Tensor. Then the user can just pass in something like `partial(torch.ones(10, 10))`.

> And generally, I wonder whether there's a similar concept to TensorInitParams somewhere in core already. There's definitely TensorOptions in C++, but I think it's just kwargs in python.

Yes, this is something I was looking into too. As you mention, I only found TensorOptions in C++.

Reply:
I looked into this and the tricky part is as follows:

1. We need the TensorProperties field since those actually represent the TensorProperties of the ShardedTensor too and need to be returned as metadata.
2. If we just change `create_op` to a Callable, we can't just call this callable with the rest arguments being `kwargs`. The reason being that especially `torch.full` and `torch.empty` have different parameters.
3. As a result, we would still need some sort if if-else block where we do comparisons like `create_op == torch.ones` etc.